metricbeat/helper/prometheus/prometheus.go

```diff
@@ -76,110 +86,224 @@ func (p *prometheus) GetFamilies() ([]*dto.MetricFamily, error) {
+			sliceSize = maxMetricsSliceSize
+		}
 
-			// Ignore retrieval errors (bad conf)
-			if value == nil {
-				continue
+		for i := 0; i < len(familyMetrics); i += sliceSize {
+			var slice []*dto.Metric
+			if i+sliceSize < len(familyMetrics) {
+				slice = familyMetrics[i : i+sliceSize]
+			} else {
+				slice = familyMetrics[i:]
 			}
+			wg.Add(1)
 
-			// Apply extra options
-			allLabels := getLabels(metric)
-			for _, option := range m.GetOptions() {
-				field, value, allLabels = option.Process(field, value, allLabels)
-			}
+			go func(idMetric int, metrics []*dto.Metric) {
+				for _, metric := range metrics {
+					m, ok := mapping.Metrics[family.GetName()]
 
-			// Convert labels
-			labels := common.MapStr{}
-			keyLabels := common.MapStr{}
-			for k, v := range allLabels {
-				if l, ok := mapping.Labels[k]; ok {
-					if l.IsKey() {
-						keyLabels.Put(l.GetField(), v)
-					} else {
-						labels.Put(l.GetField(), v)
+					// Ignore unknown metrics
+					if !ok {
+						continue
 					}
-				}
-			}
 
-			// Keep a info document if it's an infoMetric
-			if _, ok = m.(*infoMetric); ok {
-				labels.DeepUpdate(keyLabels)
-				infoMetrics = append(infoMetrics, &infoMetricData{
-					Labels: keyLabels,
-					Meta:   labels,
-				})
-				continue
-			}
+					field := m.GetField()
+					value := m.GetValue(metric)
+
+					// Ignore retrieval errors (bad conf)
+					if value == nil {
+						continue
+					}
+
+					// Apply extra options
+					allLabels := getLabels(metric)
+					for _, option := range m.GetOptions() {
+						field, value, allLabels = option.Process(field, value, allLabels)
+					}
+
+					primaryKeyLabels, secondaryKeyLabels, labels := getLabelsByType(mapping, allLabels)
+
+					if field != "" {
+						// Put it in the event if it's a common metric
+						_, event := eventsMap.getOrCreateEvent(primaryKeyLabels, secondaryKeyLabels)
+						eventsMap.updateEvent(event, field, value, labels)
+					}
+				}
+				wg.Done()
+			}(i, slice)
 
-			if field != "" {
-				// Put it in the event if it's a common metric
-				event := getEvent(eventsMap, keyLabels)
-				event.Put(field, value)
-				event.DeepUpdate(labels)
-			}
 		}
+		wg.Wait()
 	}
+}
 
-	// populate events array from values in eventsMap
-	events := make([]common.MapStr, 0, len(eventsMap))
-	for _, event := range eventsMap {
-		// Add extra fields
-		for k, v := range mapping.ExtraFields {
-			event[k] = v
+func processInfoMetrics(eventsMap *eventsMaps, mapping *MetricsMapping, families []*dto.MetricFamily) {
+	infoMetrics := []*infoMetricData{}
+	var infoMetricsMutex = &sync.Mutex{}
+
+	for _, family := range families {
+		var wg sync.WaitGroup
+
+		if _, ok := mapping.Metrics[family.GetName()]; ok {
+			continue
 		}
-		events = append(events, event)
 
+		familyMetrics := family.GetMetric()
+		sliceSize := len(familyMetrics)
+		if sliceSize > maxMetricsSliceSize {
+			sliceSize = maxMetricsSliceSize
+		}
+
+		for i := 0; i < len(familyMetrics); i += sliceSize {
```

Comment:
this loop is very similar to the one in ProcessMetrics(). can it be factored out somehow? maybe using golang version of function pointers?

Reply:
Done. Made it into one common func processFamilyMetrics()